Add cryptdev state/module for manipulating /etc/crypttab and using cryptsetup

alexander-bauer opened the pull request:
### What does this PR do?

Adds a `cryptdev` states and module for manipulating the `/etc/crypttab` file and issuing certain `cryptsetup `commands.

### What issues does this PR fix or reference?

This addresses the feature request #35816, for such a state to manipulate `/etc/crypttab`.

### Tests written?

No

### Issues to address

- [x] <del>I was not able to run the linter on my system -- I have some dependency problems that I need to look further into.</del> Fixed the linting issues
- [ ] I was not able to successfully build the HTML documentation, for the same reason
- [x] <del>I have *not* included a `versionadded` in the documentation</del> Included in c66ce0f18cb6cf07bb9de2ccf592f754087770f6

alexander-bauer committed:
Add skeleton cryptdev module

See #35816

alexander-bauer committed:
Add crypttab parsing to cryptdev module (#35816)

alexander-bauer committed:
Add crypttab deletion to cryptdev module (#35816)

alexander-bauer committed:
Add crypttab setting to cryptdev module (#35816)

alexander-bauer committed:
Clarify cryptdev.rm_crypttab documentation

See #35816

alexander-bauer committed:
Resolve missing exception import in cryptdev

See #35816

alexander-bauer committed:
List active mapped crypt devices with `cryptdev.active`

See #35816

alexander-bauer committed:
Write state for mapping crypt devices

- Add `cryptdev.mapped` for opening crypt devices and writing persistent
  `crypttab` entries.

See #35816

alexander-bauer committed:
Clarify crypttab removal module function

- `'change'` or `'absent'` is returned by `rm_crypttab` to indicate
  whether a change was made to the config file or not

See #35816

alexander-bauer committed:
Add state to unmap a crypt device

- Adds `cryptdev.unmapped` state
- Adds `cryptdev.close` module

See #35816

alexander-bauer committed:
Add `none` password behavior to editing crypttab

- When a password/keyfile is not given, the crypttab reflects `none`,
  meaning that the password must be asked at boot.

See #35816

alexander-bauer committed:
Allow crypttab to take empty options

See #35816

alexander-bauer committed:
Change cryptdev.mapping `immediate` map option

- Change `delay_mapping` with default `True` to `immediate` with default
  `False`. This is consistent with `unmapping`.

See #35816

alexander-bauer committed:
Fix mistaken up string formatting preventing immediate mapping

See #35816

alexander-bauer committed:
Require keyfile for immediate opening

See #35816

alexander-bauer committed:
Demand keyfile when mapping a cryptdevice immediately

- Prevents ungraceful error or hanging when `cryptsetup` asks for a
  password interactively
- `cryptdev.mapping` states which persist and map immediately succeed
  partially when `keyfile` is `None`, because the crypttab can be updated,
  but the immediate mapping cannot be performed
- `password` arguments in the `cryptdev` state are changed to `keyfile`,
  for clarity

See #35816

alexander-bauer committed:
Require cryptdevice is mapped either persistently or immediately

- `cryptdev.mapping` fails if neither `persist` nor `immediate` is set

See #35816

alexander-bauer committed:
Report success correctly from cryptdev.open

- Reports success and failure correctly from cryptdev.open and
  cryptdev.close

alexander-bauer committed:
Fix linting style errors for PR #41321

cachedout commented on salt/modules/cryptdev.py:
```diff
@@ -0,0 +1,344 @@
+        config='/etc/crypttab',
+        test=False,
+        match_on='name',
+        **kwargs):
```
We don't normally like to have `**kwargs` accepted for execution modules since it's impossible for the user to know what can or should be passed in here. Could you clarify why you've added them to this function, please?

alexander-bauer replied:
Thanks for your speedy review! I think I cargo-culted this accidentally from the `mount` module, on which most of this code was based. I'll remove it.

cachedout commented on salt/modules/cryptdev.py:
```diff
@@ -0,0 +1,344 @@
+
+# Define the module's virtual name
+__virtualname__ = 'cryptdev'
+
```
As you mention in the PR body itself, we do need a `versionadded` here. It will be `Oxygen`

alexander-bauer replied:
I've added the `versionadded` tag in the state and in the module in the most recent commit!

cachedout reviewed (commented):
This is really nice code. Well done!

A few minor comments in-line.

alexander-bauer committed:
Remove unnecessary **kwargs in set_crypttab

This was cargo-culted from the `mount` module, which includes it for the
sake of `salt.utils.test_mode()`, which checks all variations on
capitalization for `test` in `**kwargs` for truthiness. I have removed
the `**kwargs`, and simply check truthiness of `test` now.

See #41321

alexander-bauer committed:
Include versionadded tag for cryptdev

See #41321
